### PR TITLE
[A2-882] authz-service: fix v1 to v2 policy migration issues with applications

### PR DESCRIPTION
**Pull request by srenatus**
### :nut_and_bolt: Description

- [X] skip the v1 default policy for `service_groups`: it's implemented by additions to the `viewer` and `editor` roles in v2 (v2.1)
- [X] add an integration test to ensure we don't introduce "unknown well-known" (🤦‍♂) v1 policies again
- [x] add the required mapping logic to have customers' v1 policies for `service_groups` properly migrated
- [X] add some docs indicating how to properly add new default policies to A2

### :+1: Definition of Done

- Three ✅ above.

### :athletic_shoe: Demo Script / Repro Steps

- `rebuild components/authz-service`
- `chef-automate iam reset-to-v1`
- `export TOK=$(chef-automate admin-token)`
- create a custom `service_groups` v1 policy:
   ```
   curl -kH "api-token: $TOK" https://localhost/api/v0/auth/policies -d '{"action":"*","subjects" ["user:*"],"resource":"service_groups","effect":"allow"}'
   ```
- upgrade to v2:
```
# chef-automate iam upgrade-to-v2

Upgrading to IAM v2
Migrating v1 policies...
Creating default teams Editors and Viewers...
Skipped: Editors team already exists
Skipped: Viewers team already exists

Migrating existing teams...

Success: Enabled IAM v2
```
☝️ notice that there's none of these errors anymore:
```
Skipped policies:
  convert v1 policy "aee14d59-da0b-4974-ba6d-1a018b024874"
  unknown "well-known" policy

  convert v1 policy "9afdf13e-74ab-4797-8a31-32ece05938b4"
  could not derive v2 resource: did not recognize base v1 resource term: service_groups
```

### :chains: Related Resources

- [A2-882](https://chefio.atlassian.net/browse/A2-882)
- #462 

### :white_check_mark: Checklist

- [X] Necessary tests added/updated?
- [x] Necessary docs added/updated?
- [X] Code actually executed?
- [X] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/authz-service/server/v2/migration.go [hunk 1]
```diff
@@ -419,133 +420,127 @@ func customPolicyFromV1(pol *storage_v1.Policy) (*storage.Policy, error) {
 
 // Basically implements "Current Resource" -> "New Resource Names (RFR)" column of
 // https://docs.google.com/spreadsheets/d/1ccaYDJdMnHBfFgmNC1n2_S1YOnMJ-OgkYd8ySbb-mS0/edit#gid=363200100
-func convertV1Resource(resource string) ([]string, error) {
```
Comment (srenatus) [inline components/authz-service/server/v2/migration.go]: Drive-by: this function's return value was never more than a slice-of-one. So, I've made it a string, and fixed the fallout.

File: components/authz-service/server/v2/migration.go [hunk 2]
```diff
@@ -419,133 +420,127 @@ func customPolicyFromV1(pol *storage_v1.Policy) (*storage.Policy, error) {
...
-		terms = append([]string{"secrets"}, terms...)
-		return combineTermsIntoResourceSlice(terms), nil
+		return combineTermsIntoResource(prefixTerms("secrets", terms)...), nil
 	case "telemetry":
 		// either telemetry:config or telemetry:* maps to system:config
-		return combineTermsIntoResourceSlice([]string{"system", "config"}), nil
+		return "system:config", nil
 	case "notifications":
-		return combineTermsIntoResourceSlice(terms), nil
-	case "*":
```
Comment (srenatus) [inline components/authz-service/server/v2/migration.go]: This was a bug: `*:foo:bar` would have been converted to `*`. It's now [tighter, above](https://github.com/chef/automate/pull/612/files#diff-a9d9a31aed7e79f7b7936b91698800ceR429)

File: components/authz-service/server/v2/migration_internal_test.go [hunk 3]
```diff
@@ -2239,6 +2239,21 @@ func TestV1PolicyMigration(t *testing.T) {
 				}),
 			),
 		},
+		"service_groups,list": {
```
Comment (srenatus) [inline components/authz-service/server/v2/migration_internal_test.go]: ℹ️ Converting a custom v1 policy for service groups into a custom v2 policy.

File: components/authz-service/server/v2/migration_internal_test.go [hunk 4]
```diff
@@ -2279,6 +2294,10 @@ func TestV1PolicyMigration(t *testing.T) {
 			wellknown(t, constants_v1.CSNginxComplianceDataCollectorPolicyID),
 			checks(isSkipped()),
 		},
+		"v1 default application policy": {
```
Comment (srenatus) [inline components/authz-service/server/v2/migration_internal_test.go]: ℹ️ Skipping the default v1 application policy: it's handled by `viewer` and `editor` roles in v2.

File: integration/tests/iam_v2.sh [hunk 5]
```diff
@@ -32,14 +32,25 @@ hab_curl() {
...
       https://localhost/api/v0/auth/policies
 
     log_info "run chef-automate iam upgrade-to-v2"
-    chef-automate iam upgrade-to-v2 || return 1
+    if ! output=$(chef-automate iam upgrade-to-v2); then
+        log_error "Non-zero exit code, output:"
+        log_error "$output"
+        return 1
+    else
+        # in CI, we don't want any policies to be skipped
```
Comment (srenatus) [inline integration/tests/iam_v2.sh]: This will catch _"unknown well-known"_ policies creeping in again in the future. I've ensured that the test works by [running it on master in this build](https://buildkite.com/chef/chef-automate-master-verify-private/builds/2168#72f4fcc7-43f4-4022-a8ea-ab5d09c1ff80/578-581)

File: components/authz-service/README.md [hunk 6]
```diff
@@ -615,6 +615,18 @@ For others, it will be necessary to add and remove policies
...
+#### Adding New Default Policies
+
+To add a new default policy, the following is needed:
+
+1. A new IAM v1 policy: these are defined in [migrations](storage/postgres/migration/sql/),
+   and the [v1 constants package](constants/v1/constants.go#L3-L4).
+2. A new IAM v2 policy: they are defined in [server/v2/system](server/v2/system.go#L31).
+3. Migration logic: When upgrading from IAM v1 to v2 (v2.1), v1 policies are
+   converted. The conversion logic needs to be made aware of the new v1 policies, and
+   how (and if) they are to be migrated (if they hadn't been deleted). The procedure for
```
Comment (blakestier) [inline components/authz-service/README.md]: I think we want "if they haven't been deleted" . . . but that is confusing, also. You mean if they haven't been migrated?

File: components/authz-service/README.md [hunk 7]
```diff
@@ -615,6 +615,18 @@ For others, it will be necessary to add and remove policies
 to lock the system down so that known users are authorized to access
 specific resources.
 
+#### Adding New Default Policies
```
Comment (blakestier) [inline components/authz-service/README.md]: 👏 👏 👏

File: components/authz-service/README.md (hunk 6 above)
Comment (srenatus) [reply components/authz-service/README.md]: You're right about the grammar, for course -- what this was trying to describe is that the migration code only migrates v1 policies that haven't been deleted. There used to be shipped-with-A2 v1 policies that the user could delete. Not all "default policies" were non-deletable.

File: components/authz-service/server/v2/migration.go [hunk 8]
```diff
@@ -419,133 +420,127 @@ func customPolicyFromV1(pol *storage_v1.Policy) (*storage.Policy, error) {
...
 // https://docs.google.com/spreadsheets/d/1ccaYDJdMnHBfFgmNC1n2_S1YOnMJ-OgkYd8ySbb-mS0/edit#gid=363200100
-func convertV1Resource(resource string) ([]string, error) {
+func convertV1Resource(resource string) (string, error) {
 	terms := strings.Split(resource, ":")
 	if len(terms) == 0 {
-		return nil, errors.New("there was no resource passed")
+		return "", errors.New("there was no resource passed")
+	}
+
+	if len(terms) == 1 && terms[0] == "*" {
```
Comment (blakestier) [inline components/authz-service/server/v2/migration.go]: 💃

File: components/authz-service/server/v2/migration_internal_test.go (hunk 3 above)
Comment (blakestier) [reply components/authz-service/server/v2/migration_internal_test.go]: 🏆

File: components/authz-service/server/v2/migration_internal_test.go (hunk 4 above)
Comment (blakestier) [reply components/authz-service/server/v2/migration_internal_test.go]: 💡

File: integration/tests/iam_v2.sh [hunk 9]
```diff
@@ -32,14 +32,25 @@ hab_curl() {
...
-    local token
+    local token output
     token=$(chef-automate admin-token)
     hab_curl --fail -s -k -H "api-token: $token" \
       -d "$(jo subjects="$(jo -a user:local:alice team:local:ops)" action=read resource=auth:users)" \
       https://localhost/api/v0/auth/policies
 
     log_info "run chef-automate iam upgrade-to-v2"
-    chef-automate iam upgrade-to-v2 || return 1
+    if ! output=$(chef-automate iam upgrade-to-v2); then
```
Comment (blakestier) [inline integration/tests/iam_v2.sh]: Will this also catch `chef-automate iam upgrade-to-v2 --beta2.1`? (since we could possibly go from v1 -> v2.1, we'd need that)

File: components/authz-service/README.md [hunk 10]
```diff
@@ -615,6 +615,18 @@ For others, it will be necessary to add and remove policies
...
+
+To add a new default policy, the following is needed:
+
+1. A new IAM v1 policy: these are defined in [migrations](storage/postgres/migration/sql/),
+   and the [v1 constants package](constants/v1/constants.go#L3-L4).
+2. A new IAM v2 policy: they are defined in [server/v2/system](server/v2/system.go#L31).
```
Comment (tylercloke) [inline components/authz-service/README.md]: We should have them use the `datamigrations` for new V2 policies, otherwise they won't be applied to existing V2 installations.

File: integration/tests/iam_v2.sh (hunk 9 above)
Comment (srenatus) [reply integration/tests/iam_v2.sh]: Don't think so -- it's the same migration code.

File: components/authz-service/server/v2/migration_internal_test.go (hunk 4 above)
Comment (msorens) [reply components/authz-service/server/v2/migration_internal_test.go]: I raised the question to Kyleen, to whom this actually matters :grin:, and she has confirmed it is *not* the same. What they want is "all users to have default access to all [application] things". Right now anyone with a viewer role or on the viewers team has access, but all users are not viewers _by default_. So I think we need a migration for this policy. See https://chefio.slack.com/archives/CFCE1NQ9W/p1560797578017700

Comment (srenatus) [reply components/authz-service/server/v2/migration_internal_test.go]: I'd like to pull in Susan's feedback here. In IAM v2, we don't really follow any concept of "all users having default access" anymore.

Also, if we encourage our customers _not to migrate v1 policies_, we cannot rely on them for any nice out-of-the-box experience when using IAM v2.

File: components/authz-service/server/v2/migration.go [hunk 11]
```diff
@@ -419,133 +420,127 @@ func customPolicyFromV1(pol *storage_v1.Policy) (*storage.Policy, error) {
...
-		return nil, fmt.Errorf("did not recognize base v1 resource term: %s", terms[0])
+		return "", fmt.Errorf("did not recognize base v1 resource term: %s", terms[0])
 	}
 }
 
-func convertV1Cfgmgmt(terms []string) ([]string, error) {
+func convertV1Cfgmgmt(terms []string) (string, error) {
 	if terms[1] == "stats" {
-		return combineTermsIntoResourceSlice([]string{"infra", "nodes"}), nil
+		return "infra:nodes", nil
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authz-service/server/v2/migration.go]: Yeah, but I've opted for keeping this simple. No need to overthink it, the colons aren't going to change in IAM v2. 😉
